src/topp/Resampler.cpp

```diff
@@ -114,22 +121,84 @@ class TOPPResampler :
     //----------------------------------------------------------------
     String in = getStringOption_("in");
     String out = getStringOption_("out");
+    double sampling_rate = getDoubleOption_("sampling_rate");
+    double min_int_cutoff = getDoubleOption_("min_int_cutoff");
+    bool align_sampling = getFlag_("align_sampling");
+    bool ppm = getFlag_("ppm");
     MSExperiment<> exp;
     MzMLFile f;
     f.setLogType(log_type_);
     f.load(in, exp);
 
-    double sampling_rate = getDoubleOption_("sampling_rate");
-
-    LinearResampler lin_resampler;
     Param resampler_param;
     resampler_param.setValue("spacing", sampling_rate);
-    lin_resampler.setParameters(resampler_param);
+    if (ppm) resampler_param.setValue("ppm", "true");
+    else resampler_param.setValue("ppm", "false");
+
+    if (!align_sampling)
+    {
+      LinearResampler lin_resampler;
+      lin_resampler.setParameters(resampler_param);
+
+      // resample every scan
+      for (Size i = 0; i < exp.size(); ++i)
+      {
+        lin_resampler.raster(exp[i]);
+      }
+    }
+    else
+    {
+      LinearResamplerAlign lin_resampler;
+      lin_resampler.setParameters(resampler_param);
+
+      bool start_pos_set = false;
+      bool end_pos_set = false;
+      double start_pos = 0.0;
+      double end_pos = 0.0;
+      // get max / min positions across whole map
+      for (Size i = 0; i < exp.size(); ++i)
+      {
+        if (!exp[i].empty() && (!start_pos_set || exp[i][0].getMZ() < start_pos) )
+        {
+          start_pos = exp[i][0].getMZ();
+          start_pos_set = true;
+        }
+        if (!exp[i].empty() && (!end_pos_set || exp[i][exp[i].size()-1].getMZ() > end_pos) )
+        {
+          end_pos = exp[i][exp[i].size()-1].getMZ();
+          end_pos_set = true;
+        }
+      }
+
+      if (start_pos_set)
+      {
+        // start with even position
+        start_pos = std::floor(start_pos);
+
+        // resample every scan
+        for (Size i = 0; i < exp.size(); ++i)
+        {
+          lin_resampler.raster_align(exp[i], start_pos, end_pos);
+        }
+      }
+    }
 
-    // resample every scan
-    for (Size i = 0; i < exp.size(); ++i)
+    if (min_int_cutoff >= 0.0)
     {
-      lin_resampler.raster(exp[i]);
+      for (Size i = 0; i < exp.size(); ++i)
+      {
+        MSSpectrum<> tmp = exp[i];
+        tmp.clear(false);
+        for (Size j = 0; j < exp[i].size(); j++)
+        {
+          if (exp[i][j].getIntensity() > min_int_cutoff)
```

Comment:
actually, I was thinking that removing zeros is such a common usecase that setting cutoff=0 should be a good way of removing zeros and not cutoff=0.00001 or something that may have unwanted consequences (this may also remove small, nonzero entries). I think using a negative cutoff is a better option here if you want to keep the zeros.